### ISSUE TITLE
Kubelet to understand pods, and to be able to pull from apiserver

### ISSUE THREAD

**Issue by erictune** [GoogleCloudPlatform/kubernetes#2483]
Several inter-related goals:
 1. Kubelet continues to be usable in isolation, as well as in a Kubernetes cluster. *(composability)*
 1. Config defined in terms of Pods should work both with an apiserver, and directly with Kubelets *(bootstraping, debugging, composability)*
 1. Kubelet's can only see pods that they need to see *(security)*
 1. Don't create more apiserver API objects than necessary *(ease of understanding)*
 1. Node etcd need not be connected to master etcd. *(security, isolation, scalability)*
 1. Not have clients (be they kubectl, or kubelets or whatever) depend on the storage layer of the apiserver. *(abstraction, allowing multiple implementations of the API)*

Current state:
 1. Kubelet can read container manifest from a local file.  Or from an URL.  Used by [Google ContainerVM](https://cloud.google.com/compute/docs/containers/container_vms).
 1. Kubelets can read boundPods from their local etcd, which is connected to the master machine's etcd.
 1. Kubelets can talk to the apiserver, and do write /events.  They don't read /pods or /boundPods.

Suggested changes:
 1. Kubelet learns the definition of type Pods. 
 1. Kubelet can read json pod definitions from local files or URLs. 
    -  It also still supports containerManifests for the foreseeable future, with some way for it to determine which type to expect from a source.
 1. In a "typical" Kubernetes cluster, a Kubelet watches /api/v1beta3/pods for what pods it should run. 
 1. Get rid of BoundPod and BoundPods since nothing reads them anymore.
 1. Add a Host field to the PodStatus (but not PodSpec).  
 1. When scheduler writes a Binding, the PodStatus.Host is be set and resourceVersion is updated.  
 1. A cluster is bootstrapped by first starting one or more VMs with, 
 
Concerns that may be raised and responses:

**Q1:** Are changes to the set of pods bound made by the scheduler machine atomic or eventually consistent?

**A1:** It could work either way.  If we want atomic behavior, we could implement that in apiserver more readily that we could when we directly expose our storage via etcd.

**Q2:** Should the kubelet be allowed to see CurrentState (now PodStatus in v1beta3)?  It generates (some of) the status, so why let it see that?

**A2:** We could implement this if it is important.  Kubelet would watch pods with a selector that matches only pods with PodStatus.Host == kubelets hostname, and could use a field selector so that only PodSpec and not PodStatus is returned.

**Q3:** How do we prevent Kubelets from seeing other node's pods.
**A3:** There are a couple of ways I can think to do this with small changes to our current authorization policy.   
  1. One way is to have a distinguished "kubelet" user and special case its authorization.  
  1. Another is to create a new policy as each "kubelet" is added which matches on the SourceIP of the request, and requires a selector to be part of the request which selects on "PodStatus.Host=$SourceIP".    This may make some assumptions about the network security of the cluster, but seems like it could work.
  1. A variation on the previous is to only have one line of policy for all kubelets, but have a "condition" field in the policy that checks that PodStatus.Host matches SourceIP.
  1. Another variation is to have a different token for each kubelet and a separate policy for each.

**Comment by a-robinson** [comment]
Is there anything I can do to help move this effort along? I can take on the kubelet work if no one's started yet.

**Comment by smarterclayton** [comment]
The plan at the moment is to introduce #846, then with v1beta3 make another step.  We need to reach closure on the issues I mentioned above, and Eric had some of those elements as a WIP

**Comment by erictune** [comment]
I plan to be back working on this on Monday.

**Comment by erictune** [comment]
On GCE with kube-up.sh, kubelet and kube-proxy no longer use etcd.

**Comment by erictune** [comment]
Remaining work includes:
  - [ ] Convert more Getting-started-guides to not use etcd from the node (also may require distributing secrets to nodes)
  - [ ] Rip out etcd config source form kubelet and kube-proxy and related flags and config and code.
  - [ ] rip out boundPods from registry (keep bindings endpoint)
  - [ ] convert internal use of BoundPods to Pods in kubelet
  - [ ] optional: Kubelet can read json pod definitions from local files or URLs.

**Comment by smarterclayton** [comment]
We should note to consumers the third bullet will require you to recreate your cluster and/or migrate

**Comment by erictune** [comment]
Customers can keep running their current version of master, node daemons, and kubectl as long as they want.  

When a customer wants to upgrade, the customer ought not upgrade one component (e.g the master) to a version which is substantially different from the other components (e.g. kubelet).  

We should define what "substantially different version" is.  Maybe +/- 0.1.0.

**Comment by erictune** [comment]
Filed #3560 about how long to support things.

**Comment by bgrant0607** [comment]
Some additional thoughts:

Kubelet will need to be able to run api validation, defaulting (e.g., pullPolicy), and conversion code.

I'd eventually like to be able to point kubectl at a kubelet (if authorized). 

A pod started directly on a Kubelet should be basically the same as one started via the apisever.

**Comment by dchen1107** [comment]
I believe the lastest kubelet already run validation, but without conversion code. I made a similar comment for #3584

cc/ @thockin

**Comment by jdef** [comment]
I'd like for the scheduler to be able to add metadata to a "bound pod", however that concept ends up being realized. Currently the etcd registry ApplyBinding() implementation drops annotations on the floor.

Has there been any discussion about how a scheduler can inject annotations into a bound pod?

**Comment by smarterclayton** [comment]
Hasn't should be.  The annotations also stripped out in the Kubelet right now before the Kubelet even sees them.

----- Original Message -----

**Comment by bgrant0607** [comment]
@jdef Into the pod? Into the containers of the pod? That's relevant to #386.

My objective is for the full pod to be made available to kubelet.

**Comment by jdef** [comment]
@bgrant0607 into the pod "binding", not into the containers. If Binding and BoundPod are going away, then this means storing the binding annotations in the Pod somewhere.

Use case: a scheduler wants to add metadata into the binding that may be used on failover. For example, a recovering/restarted scheduler could query the existing "bindings" (in whatever form k8s decides to store them) and extract previously stored annotations to help rebuild internal state that was lost when the previous scheduler instance crashed. It would be very convenient to transparently leverage the existing state abstraction (registry) that kubernetes already provides.

As per the description in the *Suggested Changes* above:

> When scheduler writes a Binding, the PodStatus.Host is be set and resourceVersion is updated.

It would be awesome if the annotations were stored atomically with the binding, as part of the etcd registry implementation:

"When the scheduler writes a Binding, the PodStatus.Host is set, **additional binding annotations are saved**, and the resourceVersion is updated."

**Comment by erictune** [comment]
Binding is not going away but is currently write only.  Proposal to make
read write could be considered.

**Comment by jdef** [comment]
OK, so Binding is staying but BoundPod is going away.

What do you think about storing binding-related annotations in a new field: Pod.Status.Annotations?
* Whenever Pod.Status.Host is cleared, binding annotations are cleared
* When setting Pod.Status.Host="{something}" via assignPod(), binding annotations are copied in from Binding
* When cloning Pod.Status.Host in UpdatePod(), the Pod.Status.Annotations are also cloned

I don't think /bindings has to be a readable endpoint, as long as the binding metadata is exposed some other way. I like having changes to the binding annotations closely track changes to the Host: Pod.Status.Host seems relatively immutable, except at binding (and delete) time -- binding annotations could be the same.

**Comment by smarterclayton** [comment]
Can we simply namespace the annotations onto the pod?

**Comment by bgrant0607** [comment]
Note, this is really off topic for this issue. However...

@smarterclayton Not sure what you mean "namespace onto". I would namespace annotations attached via an automation component.

@jdef I'm not opposed to this idea, though I'd probably want to distinguish annotations on the Binding vs. those added to the pod, but do you have a concrete example of the information you'd want to record in pod annotations? If sent via Binding, I'm not sure exactly how they'd be of use to the scheduler later. By definition, once scheduled, the scheduler is done with that pod.

BTW, once set, Pod.Status.Host will never be cleared. Pods are created, scheduled, and terminate. See #3949 for more details.

**Comment by smarterclayton** [comment]
> On Feb 3, 2015, at 8:17 PM, Brian Grant <notifications@github.com> wrote:
> 
> Note, this is really off topic for this issue. However...
> 
> @smarterclayton Not sure what you mean "namespace onto". I would namespace annotations attached via an automation component.
> 
Binding annotation "foo=bar" -> pod annotation "binding/foo=bar"

**Comment by jdef** [comment]
@bgrant0607 I'm thinking about how to implement mesos task reconciliation in the kubernetes-mesos scheduler. If it crashes and then restarts, I want to be able to recover things like taskID, slaveID, offerID, etc. so as to rebuild the internal state of the scheduler. Once state is recovered, I can reconcile the against the tasks that the mesos master knows about.

If I can't use binding annotations for this, then I need to store state some other (super inconvenient) way.

Pod.Status.Host *should* never be cleared, but it looks like there's still a hack that performs exactly that in some error handler:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/registry/etcd/etcd.go#L226

**Comment by jdef** [comment]
@smarterclayton the "namespacing" approach would work for me, though it seems like it would be more effort to roll-back a merge of annotation bindings if an error strikes. Also, what if a binding k/v pair is named "alpha/beta=gamma" which would translate to "binding/alpha/beta=gamma" -- is that even allowed?

**Comment by bgrant0607** [comment]
@jdef Thanks for the use case. Yes, I agree we should support such annotations. Please file a separate issue with the example copy/pasted. As for namespaces, I think we should allow the client to specify the label namespaces. We should strongly encourage such clients to be well behaved. Since this would be a cluster service, it doesn't seem like an unreasonable imposition/assumption.

The hack you mention should be resolved by this issue, since the failure mode will be eliminated when we eliminate BoundPods.

**Comment by smarterclayton** [comment]
How are atomic scheduling constraints going to be modeled?

**Comment by lavalamp** [comment]
kubelet will do the constraint check(s) and reject offending pods.

**Comment by smarterclayton** [comment]
So the status would be updated to "rejected" or similar?  Is rejection temporary or permanent (ie should rcs give up immediately and delete the pod)?

**Comment by erictune** [comment]
rcs should retry, but maybe back off on the rate after a while.

If the is on a single node with an apiserver-based,.

If the port conflict is with an existing apiserver-based pod, then the
scheduler will probably notice the node conflict next time and find a
different node that works (or leave it pending if there is no node that
works).

If the port conflict is with an existing file-based pod, which the
schedulers don't know about (yet), then the behavior is also similar to
today, but we will be able to fix it once file-based pods are surfaced
(Read-only) to the apiserver.

**Comment by bgrant0607** [comment]
Similar to discussion re. out of disk, OOM, host port conflict, etc., the pod should just be failed, with a specific reason. 

Whenever a pod's Spec.Host field is set (at creation time or during binding), the fit predicates should be tested. The update should fail if they don't pass at that time, as a best-effort check. At least policy-style checks should be applied (e.g., sole tenancy).

However, there's no significant difference between a bad scheduling decision based on stale data and a change in conditions on the host shortly after the decision is made that results in failure. The scheduler needs to be resilient to such failures. There are an unbounded number of reasons why they can happen. For instance, hosts can be broken in subtle ways that only affect containers using certain devices or other system resources or features (e.g., perhaps pulls of new images could be broken but existing images would work). New pods could be started directly on the Kubelets. Existing containers could launch new ones, or otherwise start to consume a lot more resources. Currently we don't have differentiated quality of service, so nothing stops launching new pods that don't specify resource limits (unless using the admission controller, but even then we don't have cpu or disk enforcement). Containers/processes that are supposed to be dead could be unkillable, preventing reuse of their resources. The kernel could consume an unexpectedly large amount of memory and never give it back. The host could disappear. On at least physical machines, individual (non-root) disks could fail. 2 containers could massively conflict in the cache, reducing performance by 10x.

Omega implemented scheduling atomicity at the storage level. It's not worth it.

We're also far from being in a position where it would be important. For example, we're neither replicating apiserver nor running multiple schedulers at the moment.

Just as the node controller kills pods on nodes that are unresponsive, it could (reactively) remove pods from overcommitted nodes.

Eventually there will be a lot of churn, both on individual machines and in the cluster as a whole:
* horizontal auto-scaling
* vertical auto-sizing
* automatic pushes
* batch, CI, and workflow workloads
* automated node maintenance (e.g., kernel updates)

We're going to need a background (proactive) self-healing controller to rebalance pods, which I think of as similar to a generational garbage collector.

**Comment by erictune** [comment]
agree with everything @bgrant0607 just said

**Comment by bgrant0607** [comment]
BoundPods is hypothesized to be gating apiserver pod scheduling and deletion performance, and perhaps is causing pathologically bad resource conflicts resulting in high latency and apiserver cpu consumption. #3954, #4521.

**Comment by erictune** [comment]
Closing in favor of targeted issues:
  - Kubelet config/file should read Pod #3372
  - Kubelet to use Pods internally. #5392
